File: backend/app/services/cron.py
from datetime import datetime, timedelta
from sqlalchemy.orm import Session
from .. import models
from .youtube import fetch_weekly_uploads_count

def run_weekly_youtube_audit(db: Session, mock_override_uploads: dict = None) -> dict:
    """
    Weekly cron job running every Sunday at 23:59 (Cláusula 10ª).
    Audits uploads for all active condôminos.
    Returns audit details.
    """
    # Determine the week code, e.g., "2026-W23"
    today = datetime.utcnow()
    year, week_num, _ = today.isocalendar()
    semana_codigo = f"{year}-W{week_num:02d}"

    # YouTube API uses RFC 3339 datetime format for publishedAfter (e.g., 7 days ago)
    seven_days_ago = today - timedelta(days=7)
    published_after_str = seven_days_ago.isoformat() + "Z"

    # Get all active condôminos
    condominos = db.query(models.Condomino).all()
    results = []

    for c in condominos:
        # We only audit uploads for active adimplente condôminos
        if c.status != models.OperacionalStatus.ATIVO_ADIMPLENTE and c.status != models.OperacionalStatus.BLOQUEADO_ASSIDUIDADE:
            continue

        # Get uploads count
        # Support front-end mock overrides for testing
        if mock_override_uploads and c.id in mock_override_uploads:
            uploads_count = mock_override_uploads[c.id]
        else:
            youtube_id = c.youtube_id or f"mock_chan_{c.nome_comercial.lower()}"
            uploads_count = fetch_weekly_uploads_count(youtube_id, published_after_str)

        # Check validity (minimum 1 video)
        is_valid = uploads_count >= 1

        # Check if already has an entry for this week
        existing_entry = db.query(models.EntregaVideo).filter(
            models.EntregaVideo.condomino_id == c.id,
            models.EntregaVideo.semana_codigo == semana_codigo
        ).first()

        if existing_entry:
            existing_entry.qtd_entregue = uploads_count
            existing_entry.status_valido = is_valid
        else:
            new_entry = models.EntregaVideo(
                condomino_id=c.id,
                semana_codigo=semana_codigo,
                qtd_entregue=uploads_count,
                status_valido=is_valid
            )
            db.add(new_entry)

        # Handle State transitions (Cláusula 10ª / 11ª)
        old_status = c.status
        if not is_valid:
            c.status = models.OperacionalStatus.BLOQUEADO_ASSIDUIDADE
        else:
            # Restore to ATIVO_ADIMPLENTE if they were blocked and now uploaded
            if c.status == models.OperacionalStatus.BLOQUEADO_ASSIDUIDADE:
                c.status = models.OperacionalStatus.ATIVO_ADIMPLENTE

        results.append({
            "condomino": c.nome_comercial,
            "semana": semana_codigo,
            "qtd_entregue": uploads_count,
            "old_status": old_status,
            "new_status": c.status,
            "status_valido": is_valid
        })

    db.commit()
    return {
        "status": "success",
        "semana": semana_codigo,
        "audited_count": len(results),
        "details": results
    }

File: src/lib/services/cron.ts
import { supabase } from "../supabase";
import { fetchWeeklyUploadsCount } from "./youtube";

interface AuditResult {
  condomino: string;
  semana: string;
  qtd_entregue: number;
  old_status: string;
  new_status: string;
  status_valido: boolean;
}

export async function runWeeklyYoutubeAudit(mockOverrideUploads: Record<string, number> | null = null): Promise<any> {
  // Determine week code (e.g. 2026-W23)
  const today = new Date();
  
  // Calculate ISO week number
  const target = new Date(today.valueOf());
  const dayNr = (today.getDay() + 6) % 7;
  target.setDate(target.getDate() - dayNr + 3);
  const firstThursday = target.valueOf();
  target.setMonth(0, 1);
  if (target.getDay() !== 4) {
    target.setMonth(0, 1 + ((4 - target.getDay() + 7) % 7));
  }
  const weekNum = 1 + Math.ceil((firstThursday - target.valueOf()) / 604800000);
  const year = today.getFullYear();
  const semanaCodigo = `${year}-W${String(weekNum).padStart(2, "0")}`;

  // Time 7 days ago in ISO format
  const sevenDaysAgo = new Date(today.getTime() - 7 * 24 * 60 * 60 * 1000);
  const publishedAfterStr = sevenDaysAgo.toISOString();

  // Get all condominos
  const { data: condominos, error: fetchError } = await supabase
    .from("condominos")
    .select("*");

  if (fetchError || !condominos) {
    throw new Error(`Failed to fetch condominos for audit: ${fetchError?.message}`);
  }

  const results: AuditResult[] = [];

  for (const c of condominos) {
    // Only audit active/blocked creators
    if (c.status !== "ATIVO_ADIMPLENTE" && c.status !== "BLOQUEADO_ASSIDUIDADE") {
      continue;
    }

    let uploadsCount = 2; // Default
    if (mockOverrideUploads && mockOverrideUploads[c.id] !== undefined) {
      uploadsCount = mockOverrideUploads[c.id];
    } else {
      const youtubeId = c.youtube_id || `mock_chan_${c.nome_comercial.toLowerCase()}`;
      uploadsCount = await fetchWeeklyUploadsCount(youtubeId, publishedAfterStr);
    }

    const isValid = uploadsCount >= 1;
    const oldStatus = c.status;
    let newStatus = oldStatus;

    if (!isValid) {
      newStatus = "BLOQUEADO_ASSIDUIDADE";
    } else {
      if (oldStatus === "BLOQUEADO_ASSIDUIDADE") {
        newStatus = "ATIVO_ADIMPLENTE";
      }
    }

    // Upsert the video delivery record
    const { data: existingEntry } = await supabase
      .from("entregas_video")
      .select("*")
      .eq("condomino_id", c.id)
      .eq("semana_codigo", semanaCodigo)
      .single();

    if (existingEntry) {
      await supabase
        .from("entregas_video")
        .update({
          qtd_entregue: uploadsCount,
          status_valido: isValid,
        })
        .eq("id", existingEntry.id);
    } else {
      await supabase
        .from("entregas_video")
        .insert({
          condomino_id: c.id,
          semana_codigo: semanaCodigo,
          qtd_entregue: uploadsCount,
          status_valido: isValid,
        });
    }

    // Update status if changed
    if (newStatus !== oldStatus) {
      await supabase
        .from("condominos")
        .update({ status: newStatus })
        .eq("id", c.id);
    }

    results.push({
      condomino: c.nome_comercial,
      semana: semanaCodigo,
      qtd_entregue: uploadsCount,
      old_status: oldStatus,
      new_status: newStatus,
      status_valido: isValid,
    });
  }

  return {
    status: "success",
    semana: semanaCodigo,
    audited_count: results.length,
    details: results,
  };
}
